simpleflow/swf/executor.py

```diff
@@ -1156,34 +1158,32 @@ def propagate_signals(self):
             if not propagate:
                 continue
             name = signal['name']
-            orig_workflow_id = input.get('__workflow_id')
-            orig_run_id = input.get('__run_id')
-
-            input = {
-                'args': input.get('args'),
-                'kwargs': input.get('kwargs'),
-                '__workflow_id': self._workflow_id,
-                '__run_id': self._run_id,
-            }
+
+            args = input.get('args', ())
+            kwargs = input.get('kwargs', {})
             sender = (
-                signal['external_workflow_id'] or orig_workflow_id,
-                signal['external_run_id'] or orig_run_id
+                signal['external_workflow_id'],
+                signal['external_run_id']
             )
             signaled_workflows_ids = set(
                 (w['workflow_id'], w['run_id']) for w in history.signaled_workflows[name]
             )
-            signaled_workflows_ids.add((orig_workflow_id, orig_run_id))
             not_signaled_workflows_ids = list(known_workflows_ids - signaled_workflows_ids - {sender})
+            extra_input = {'__propagate': propagate}
             for workflow_id, run_id in not_signaled_workflows_ids:
-                try:
-                    self._execution.signal(
-                        signal_name=name,
-                        input=input,
-                        workflow_id=workflow_id,
-                        run_id=run_id,
-                    )
-                except swf.models.workflow.WorkflowExecutionDoesNotExist:
-                    logger.info('Workflow {} {} disappeared'.format(workflow_id, run_id))
+                self.schedule_task(SignalTask(
+                    name,
+                    workflow_id,
+                    run_id,
+                    None,
+                    extra_input,
+                    *args,
+                    **kwargs
+                ))
+                signals_scheduled = True
+        if signals_scheduled:
+            self._append_timer = True
```

Comment:
Note: I think `self._append_timer` is unneeded, since we raise (and so the webflow cannot end immediately): we'll have future(s) from the signal(s) that will be finished or not as intended.